Support FakeTensor with FlatParameter

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101987

In this PR we turn FlatParameter into a virtual tensor subclass
which doesn't actually ever get instantiated: __new__ will create
a Parameter instead (or a FakeTensor, if necessary).

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
Support FakeTensor with FlatParameter

In this PR we turn FlatParameter into a virtual tensor subclass
which doesn't actually ever get instantiated: __new__ will create
a Parameter instead (or a FakeTensor, if necessary).

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

awgu reviewed (approved):
This change looks good to me!

ezyang commented on test/test_fake_tensor.py:
```diff
@@ -67,6 +68,16 @@ def test_parameter_instantiation(self):
+        from torch.distributed.fsdp.flat_param import FlatParameter
+        with FakeTensorMode() as m:
+            data = torch.randn(2, 2)
+            param = FlatParameter(data, requires_grad=True)
```
Hmmm... can we table this for now? While in principle it makes sense, I'm not sure I want to directly call `from_tensor` on a FlatParameter; I think we probably want to model it specially in Dynamo, in particular.

ezyang commented:
Alas, this apparently is not semantics preserving! What a pain.

ezyang committed:
Update on "Support FakeTensor with FlatParameter"


In this PR we turn FlatParameter into a virtual tensor subclass
which doesn't actually ever get instantiated: __new__ will create
a Parameter instead (or a FakeTensor, if necessary).

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_common_utils.py:
```diff
@@ -283,7 +283,7 @@ def module_fn(module, prefix, tree_level, param_to_fqns):
             local_fqns = (
                 param._fqns
-                if type(param) is flat_param_file.FlatParameter
+                if isinstance(param, flat_param_file.FlatParameter)
```
For context, I think we had to change these to `type(param) is FlatParameter` because otherwise other tensor subclasses with `_is_param = True` return `True` from the `instance(tensor, FlatParameter)` check. See this PR description for more details: https://github.com/pytorch/pytorch/pull/85038

Possibly your new override of `__instancecheck__()` resolves that issue?

ezyang replied:
Yes it does! This is also an interesting failure mode for the `_is_param` trick from Parameter which might explain why @jbschlosser found it was BC-breaking to do the `_is_param` trick universally.

ezyang commented:
these mypy errors 🤔

ezyang committed:
Update on "Support FakeTensor with FlatParameter"


In this PR we turn FlatParameter into a virtual tensor subclass
which doesn't actually ever get instantiated: __new__ will create
a Parameter instead (or a FakeTensor, if necessary).

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Support FakeTensor with FlatParameter"


In this PR we turn FlatParameter into a virtual tensor subclass
which doesn't actually ever get instantiated: __new__ will create
a Parameter instead (or a FakeTensor, if necessary).

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

bdhirsh commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -173,7 +173,17 @@ class FlatParamShardMetadata(NamedTuple):
+class _FlatParameterMeta(torch._C._TensorMeta):
+    # Make `isinstance(t, FlatParameter)` return True for custom tensor
+    # instances that have the _is_flat_param flag for BC
+    def __instancecheck__(self, instance):
```
🤯
